File: exercise/8/solution.cpp
class Solution {
public:
    int myAtoi(string str) {
        if (str.length() == 0) {
            return 0;
        }
        int state = 0, pos = 0, sign = 1, num = 0, overflow = 0;
        while (true) {
            if (pos == str.length()) {
                state = 2;
            }
            switch (state) {
                case 0:
                    if (str[pos] == ' ') {
                        ;
                    } else if (str[pos] == '+') {
                        state = 3;
                    } else if (str[pos] == '-') {
                        sign = -1;
                        state = 3;
                    } else if (str[pos] >= '0' && str[pos] <= '9') {
                        num = str[pos] - '0';
                        state = 1;
                    } else {
                        state = 2;
                    }
                    break;
                case 3:
                    if (str[pos] >= '0' && str[pos] <= '9') {
                        num = str[pos] - '0';
                        state= 1;
                    } else {
                        state = 2;
                    }
                    break;
                case 1:
                    if (str[pos] >= '0' && str[pos] <= '9') {
                        int now = str[pos] - '0';
                        if (sign == -1) {
                            if (num == 214748364 && now >= 8 || num > 214748364) {
                                state = 2;
                                overflow = 1;
                            } else {
                                num *= 10;
                                num += now;
                            }
                        } else {
                            if (num == 214748364 && now >= 7 || num > 214748364) {
                                state = 2;
                                overflow = 1;
                            } else {
                                num *= 10;
                                num += now;
                            }
                        }
                    } else {
                        state = 2;
                    }
                    break;
                case 2:
                    if (overflow == 1) {
                        if (sign == 1) {
                            return 2147483647;
                        } else {
                            return -2147483648;
                        }
                    } else {
                        return sign * num;
                    }
            }
            pos++;
        }
    }
};

File: exercise/8/solution.py
class Solution:
    def myAtoi(self, str: str) -> int:
        num_table = {"0": 0, "1": 1, "2": 2, "3": 3, "4": 4, "5": 5, "6": 6, "7": 7, "8": 8, "9": 9}
        state = 0
        sign = 1
        num = 0
        for c in str:
            if state == 0:
                if c == ' ':
                    continue
                elif c == '+':
                    state = 1
                elif c == '-':
                    sign = -1
                    state = 1
                elif c in num_table:
                    num = num_table[c]
                    state = 1
                else:
                    break
            elif state == 1:
                if c in num_table:
                    if num > 214748365 or sign == -1 and num == 214748364 and num_table[c] > 8 or sign == 1 and num == 214748364 and num_table[c] >= 8:
                        state = 2
                        if sign == 1:
                            num = 2147483647
                        else:
                            num = 2147483648
                    else:
                        num = num * 10 + num_table[c]
                else:
                    state = 2
            elif state == 2:
                return num
        return num
